### PR TITLE
Delete old scores automatically from plugin

### LINKED ISSUE #193: Automatically removing old scores for each plugin
We can keep 5 or 10 scores for each plugins to get a trend, but we don't need to keep all the scores ever computed.
This needs to be scheduled and well tested.

- [ ] Removes scores for each plugins except the last 5 (by `computed_at`).
- [ ] Schedule this process to run every hour.
- [ ] This needs to be tested.

### PR DESCRIPTION
**Pull request by Jagrutiti**
### Description

Resolves #193.

Plugins scores, even if not computed at each run of the `ScoreEngine`, can start to fill up the scores table with new entries. 
Keeping some entries for each plugin can help users and contributors have a sense of the evolution of the plugin. But we don't need to keep all of them. 
Keeping 5 scores can be enough. This count could be changed in the future.
 

### Submitter checklist

- [x] If the issue exists, it is well described and linked in the description
- [x] The description of this pull request is detailed and explain why this pull request is needed
- [x] The changeset is on a specific branch
  - `feature/` for new feature, or improvements
  - `fix/` for bug fixes
  - `docs/` for any documentation changes
- [x] If required, the documentation has been updated
- [x] There is automated tests to cover the code change / addition
    - If there is no test, include an explanation why in the description
- [x] Run `mvn verify` locally and all tests are passing successfully
  - It is OK to create a pull request which has failing tests if it is created as a draft, is to fix a bug and the first commit is the test to prove the existence of the bug.
- [x] There is no new warnings (checkstyle nor spotbugs) on the code

### CODE REVIEW

**Comment by alecharp** [comment]
This seems really promising! Could you add tests and create a scheduled method to run this on a daily basis maybe? Thanks!

**Comment by Jagrutiti** [comment]
Hey @alecharp 

I have added the scheduler. 

I need help with the test cases. How do I test the query? Because the tables in Docker cmd do not have any data.

**Comment by alecharp** [comment]
You can persist data in your test method. You can find example of that in https://github.com/jenkins-infra/plugin-health-scoring/blob/b9133e157595e367696a390599847919f92f8b61/core/src/test/java/io/jenkins/pluginhealth/scoring/repository/PluginRepositoryIT.java#L75-L83 but also in other integration tests classes.

**Comment by alecharp** [comment]
> I have added the test case and it passes. How shall I test the scheduler?

There is no need imo to test the scheduler.

File: core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryTest.java
Comment (alecharp) [inline core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryTest.java:1]: This looks more like an integration test than a unit test. 
The class name should be suffixed with  `IT` not `Test`.

File: core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryTest.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+    @Autowired
+    private ScoreRepository repository;
+
+    @Autowired private TestEntityManager entityManager;
+
+    @Test
+    void shouldBeAbleToDeleteOldScores() {
+        final Plugin plugin1 = new Plugin("plugin-1", new VersionNumber("1.0"), "scm", ZonedDateTime.now());
+        entityManager.persist(plugin1);
```
Comment (alecharp) [inline core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryTest.java:31]: You need to assign the value returned by the `entityManager.persist` to the variable `plugin1`

```suggestion
        final Plugin plugin1 = entityManager.persist(new Plugin("plugin-1", new VersionNumber("1.0"), "scm", ZonedDateTime.now()));
```

And the same for all other persisted values.

File: core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryTest.java [hunk 2]
```diff
@@ -12,9 +22,64 @@ public class ScoreRepositoryTest {
...
+        entityManager.persist(score34);
+        final Score score35 = new Score(plugin3, ZonedDateTime.now());
+        entityManager.persist(score35);
+        final Score score36 = new Score(plugin3, ZonedDateTime.now());
+        entityManager.persist(score36);
+        final Score score37 = new Score(plugin3, ZonedDateTime.now());
+        entityManager.persist(score37);
+
         long noOfRowsDeleted = repository.deleteOldScoreFromPlugin();
+        assertThat(noOfRowsDeleted).isEqualTo(6);
```
[the remark this answers is not in the document]
Comment (alecharp) [reply core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryTest.java:83]: Yes that would be better. Please change the `timestamp` of the `scoreXX` variable to create some "span" in the timestamp. You can use `ZonedDateTime.now().minusDays(1)` or something like that.

**Review by alecharp — CHANGES_REQUESTED** [review]
Please see my comments on the proposed changeset.

**Comment by Jagrutiti** [comment]
Hey @alecharp 

I have incorporated all your suggestions.

**Comment by alecharp** [comment]
Please make sure that the correct scores were deleted, not the latest ones.

**Comment by Jagrutiti** [comment]
> Please make sure that the correct scores were deleted, not the latest ones.

Done

File: core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryIT.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        final Score recentScore001 = entityManager.persist(new Score(plugin3, ZonedDateTime.now().minusDays(5)));
+        final Score recentScore002 = entityManager.persist(new Score(plugin3, ZonedDateTime.now().minusDays(4)));
+        final Score recentScore003 = entityManager.persist(new Score(plugin3, ZonedDateTime.now().minusDays(3)));
+        final Score recentScore004 = entityManager.persist(new Score(plugin3, ZonedDateTime.now().minusDays(2)));
+        final Score recentScore005 = entityManager.persist(new Score(plugin3, ZonedDateTime.now().minusDays(1)));
+
+        long noOfRowsDeleted = repository.deleteOldScoreFromPlugin();
+        assertThat(noOfRowsDeleted).isEqualTo(6);
+
+        List<Score> remainingScores = entityManager.getEntityManager().createQuery("SELECT s FROM Score s", Score.class).getResultList();
```
Comment (alecharp) [inline core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryIT.java:62]: You could use the `findAll` method on the repository object.

**Review by alecharp — CHANGES_REQUESTED** [review]

File: core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryIT.java (hunk 3 above)
Comment (Jagrutiti) [reply core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryIT.java]: > You could use the `findAll` method on the repository object.

Done.

File: core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryIT.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import io.jenkins.pluginhealth.scoring.model.Score;
+
+import hudson.util.VersionNumber;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;
+
+
```
Comment (alecharp) [inline core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryIT.java:19]:
```suggestion
```

File: core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryIT.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
+@DataJpaTest
+public class ScoreRepositoryIT extends AbstractDBContainerTest {
+
+    @Autowired
+    private ScoreRepository repository;
+
+    @Autowired private TestEntityManager entityManager;
```
Comment (alecharp) [inline core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryIT.java:27]:
```suggestion
    @Autowired private ScoreRepository repository;
    @Autowired private TestEntityManager entityManager;
```

File: core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryIT.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        List<Score> remainingScores = repository.findAll();
+
+        assertThat(noOfRowsDeleted).isEqualTo(6);
+        assertThat(remainingScores.size()).isEqualTo(15);
+        assertThat(remainingScores).doesNotContain(oldScore1, oldScore2, oldScore21, oldScore22, oldScore31, oldScore32);
+
+        assertThat(remainingScores).containsExactlyInAnyOrder(recentScore1, recentScore2, recentScore3, recentScore4, recentScore5,
+            recentScore01, recentScore02, recentScore03, recentScore04, recentScore05,
+            recentScore001, recentScore002, recentScore003, recentScore004, recentScore005);
+
```
Comment (alecharp) [inline core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryIT.java:69]: I think having both `doesNotContain` and `containsExactlyInAnyOther` validation is redundant but better safe than sorry.

```suggestion
        assertThat(remainingScores)
            .hasSize(15)
            .doesNotContain(oldScore1, oldScore2, oldScore21, oldScore22, oldScore31, oldScore32)
            .containsExactlyInAnyOrder(
                 recentScore1, recentScore2, recentScore3, recentScore4, recentScore5,
                 recentScore01, recentScore02, recentScore03, recentScore04, recentScore05,
                 recentScore001, recentScore002, recentScore003, recentScore004, recentScore005
            );
```

File: war/src/main/java/io/jenkins/pluginhealth/scoring/schedule/DeleteOldScoreScheduler.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+import io.jenkins.pluginhealth.scoring.service.ScoreService;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Component;
+
+@Component
+public class DeleteOldScoreScheduler {
+
```
Comment (alecharp) [inline war/src/main/java/io/jenkins/pluginhealth/scoring/schedule/DeleteOldScoreScheduler.java:12]:
```suggestion
```

File: core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryIT.java
Comment (alecharp) [inline core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryIT.java:1]: Please make sure to add the license header.

File: war/src/main/java/io/jenkins/pluginhealth/scoring/schedule/DeleteOldScoreScheduler.java
Comment (alecharp) [inline war/src/main/java/io/jenkins/pluginhealth/scoring/schedule/DeleteOldScoreScheduler.java:1]: Please add the license header.

**Review by alecharp — CHANGES_REQUESTED** [review]
This looks better.

I didn't see we were missing the license header on the two new files.
Others are just "text" formatting. I'll look into improving checkstyle to help with that in the future.

File: core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryIT.java
Comment (Jagrutiti) [reply core/src/test/java/io/jenkins/pluginhealth/scoring/repository/ScoreRepositoryIT.java:1]: Done

File: war/src/main/java/io/jenkins/pluginhealth/scoring/schedule/DeleteOldScoreScheduler.java
Comment (Jagrutiti) [reply war/src/main/java/io/jenkins/pluginhealth/scoring/schedule/DeleteOldScoreScheduler.java:1]: Done

**Review by alecharp — COMMENTED** [review]
This is good for me. Thanks.

**Pull request merged**